### PR TITLE
Working version

### PR DESCRIPTION
**Pull request by Xapp73**

### CODE REVIEW

File: src/Form/UserForm.php [hunk 1]
```diff
@@ -1,227 +1,262 @@
...
+                'type' => 'password',
                 'name' => 'confirm_password',
                 'options' => [
                     'label' => 'Confirm password',
                 ],
             ]);
         }
-        
+
+        $roles = $this->entityManager->getRepository(Role::class)->findAll();
```
Comment (ikovalyov) [inline src/Form/UserForm.php]: shouldn't be here at all

File: src/Form/UserForm.php [hunk 2]
```diff
@@ -1,227 +1,262 @@
...
+        }
+
+        // Add role field selector here
+        $this->add([
+            'type' => 'select',
+            'name' => 'role',
+
+            'options' => [
+                'label' => 'Role',
+                'value_options' => $rolesselector,
```
Comment (ikovalyov) [inline src/Form/UserForm.php]: get this info in controller

File: src/Form/UserForm.php [hunk 3]
```diff
@@ -1,227 +1,262 @@
...
+        ]);
+
+        if ($this->scenario == 'create') {
+
+            // Add input for "password" field
+            $inputFilter->add([
+                'name' => 'password',
                 'required' => true,
-                'filters'  => [                    
-                    ['name' => 'StringTrim'],
```
Comment (ikovalyov) [inline src/Form/UserForm.php:166]: StringTrim filter was removed?

File: src/Service/UserManager.php [hunk 4]
```diff
@@ -38,7 +39,13 @@ public function addUser($data)
         // Create new User entity.
         $user = new User();
         $user->setEmail($data['email']);
-        $user->setFullName($data['full_name']);        
+        $user->setFullName($data['full_name']);
+
+        // Get role object based on role Id from form
+        /** @var Role $role */
+        $role = $this->entityManager->find(Role::class, ['roleId' => $data['role']]);
```
Comment (ikovalyov) [inline src/Service/UserManager.php]: can't you just pass role id here? I guess doctrine is smart enough.

File: src/Service/UserManager.php [hunk 5]
```diff
@@ -71,7 +82,13 @@ public function updateUser($user, $data)
...
         $user->setEmail($data['email']);
         $user->setFullName($data['full_name']);        
-        $user->setStatus($data['status']);        
+        $user->setStatus($data['status']);
+
+        // Get role object based on role Id from form
+        /** @var Role $role */
+        $role = $this->entityManager->find(Role::class, ['roleId' => $data['role']]);
+        // Set role to user
+        $user->addRole($role);
```
Comment (ikovalyov) [inline src/Service/UserManager.php:96]: same

Comment (ikovalyov) [reply src/Service/UserManager.php:96]: Also don't forget to remove previous role

File: src/Service/UserManager.php [hunk 6]
```diff
@@ -95,7 +112,11 @@ public function createAdminUserIfNotExists()
             $user->setPassword($passwordHash);
             $user->setStatus(User::STATUS_ACTIVE);
             $user->setDateCreated(date('Y-m-d H:i:s'));
-            
+            // Get role object based on role Id from form
+            /** @var Role $role */
+            $role = $this->entityManager->find(Role::class, ['roleId' => 2]);
+            // Set role to user
+            $user->addRole($role);
```
Comment (ikovalyov) [inline src/Service/UserManager.php:124]: same

File: src/Service/UserManager.php [hunk 7]
```diff
@@ -95,7 +112,11 @@ public function createAdminUserIfNotExists()
             $user->setPassword($passwordHash);
             $user->setStatus(User::STATUS_ACTIVE);
             $user->setDateCreated(date('Y-m-d H:i:s'));
-            
+            // Get role object based on role Id from form
+            /** @var Role $role */
+            $role = $this->entityManager->find(Role::class, ['roleId' => 2]);
```
Comment (ikovalyov) [inline src/Service/UserManager.php]: 2 should be a constant

File: view/user/user/add.phtml [hunk 8]
```diff
@@ -19,6 +19,10 @@ $form->get('full_name')->setAttributes([
     'placeholder'=>'John Doe'
     ]);
 
+$form->get('role')->setAttributes([
```
Comment (ikovalyov) [inline view/user/user/add.phtml:22]: write it in template instead of using jquery

File: view/user/user/edit.phtml [hunk 9]
```diff
@@ -19,6 +19,10 @@ $form->get('full_name')->setAttributes([
     'placeholder'=>'John Doe'
     ]);
 
+$form->get('role')->setAttributes([
```
Comment (ikovalyov) [inline view/user/user/edit.phtml:22]: same

File: view/user/user/add.phtml (hunk 8 above)
Comment (Xapp73) [reply view/user/user/add.phtml:22]: did it as it is at native code, so no idea what do You want here

File: view/user/user/edit.phtml (hunk 9 above)
Comment (Xapp73) [reply view/user/user/edit.phtml:22]: did it as it is at native code, so no idea what do You want here

File: src/Form/UserForm.php (hunk 3 above)
Comment (Xapp73) [reply src/Form/UserForm.php:166]: No, it seems that You have been mistaken by old filters and new filters location

File: view/user/user/add.phtml (hunk 8 above)
Comment (ikovalyov) [reply view/user/user/add.phtml:22]: shitty code :(

File: src/Controller/UserController.php [hunk 10]
```diff
@@ -9,334 +12,379 @@
...
         ]);
     }
+
+    /**
+     * @return mixed
+     */
+    public function getRolesSelector()
+    {
+        $roles = $this->entityManager->getRepository(Role::class)->findAll();
+        $hydrator = new \Zend\Hydrator\ClassMethods();
```
Comment (ikovalyov) [inline src/Controller/UserController.php]: use ClassMethods

File: src/Controller/UserController.php [hunk 11]
```diff
@@ -9,334 +12,379 @@
...
+
+    /**
+     * @param User $user
+     * @return int
+     */
+    public function getUserRole($user)
+    {
+        // checking for existing role if editing mode
+        $rolecurrent['role_id'] = self::GUEST_ROLE_ID;
+        $hydrator = new \Zend\Hydrator\ClassMethods();
```
Comment (ikovalyov) [inline src/Controller/UserController.php]: same

File: src/Service/AuthManager.php [hunk 12]
```diff
@@ -110,7 +124,13 @@ public function filterAccess($controllerName, $actionName)
                     $actionList=='*') {
                     if ($allow=='*')
                         return true; // Anyone is allowed to see the page.
-                    else if ($allow=='@' && $this->authService->hasIdentity()) {
+                    else if (($allow=='@' || is_array($allow)) && $this->authService->hasIdentity()) {
```
Comment (ikovalyov) [inline src/Service/AuthManager.php]: I don't really like the @ behaviour. It cab be easily resolved with "user" role which cover every logged in user.

**Pull request merged**